### ISSUE TITLE
Callback before condition is evaluated

### ISSUE THREAD

**Issue by dmitriy-yefremov** [awaitility/awaitility#53]
It would be very nice to be able to intercept the lifecycle of a condition each time before it is evaluated. This hook could be used to refresh the state before checking the condition.

I'm thinking of the following syntax:
`await("data).atMost(ONE_MINUTE).repeat(this::refreshData).until(this::checkData);`

In this example the "refreshData" method fetches some data from a remote data store and saves it in a class field. The "checkData" checks the field if it matches a certain condition.

Currently I have to combine both operations into the "checkData" method. Introducing that new concept would make my code more modular and reusable.

Thanks!

**Comment by johanhaleby** [comment]
Hi, 

Would a [ConditionEvaluationListener](https://github.com/awaitility/awaitility/wiki/Usage#condition-evaluation-listener) work for you?

**Comment by dmitriy-yefremov** [comment]
Not quite. AFAIU it is called **after** the condition is evaluated. I was thinking of something called **before**. Let's say there is some remote state that is also cached locally. I was thinking of fetching the state in repeat() and checking the local copy in until(). I can always combine the two and execute both in until(), but that may not be as clean.

```
public class Test {

  private State state;

  @Test
  public void testSomething() {
    await().atMost(ONE_MINUTE)
        .repeat(() -> state = fetchStateFromServer())
        .until(() -> state.someField().equals("expected value"));
  }

}
```

What do you think?

**Comment by johanhaleby** [comment]
I see what you mean and I can see that it could be useful. Calling `fetchStateFromServer` from within the condition complects to different things so splitting it up along the lines you suggest would make sense. 

This is not something I've ever encountered myself though (and never heard this requested before) so I just wonder if the case for this is so slim that it might warrant not a new API method. But I like the idea and I can see the usefulness of it. Need to think about it a bit :)

**Comment by dmitriy-yefremov** [comment]
This is not a show stopper obviously. Just a nice to have feature. The same code could be rewritten as follows using the existing API:

```
public class Test {

  @Test
  public void testSomething() {
    await().atMost(ONE_MINUTE)
        .until(() -> fetchStateFromServer().someField().equals("expected value"));
  }

}
```

I'm fine either way. Just wanted to share my thoughts. And BTW thanks for the great library! :-)

**Comment by johanhaleby** [comment]
Sure this is how I would usually do it. But I like the fact that your idea would decouple fetching/generation of data from the assertion. This would make it more [simple](https://www.infoq.com/presentations/Simple-Made-Easy).
